python/triton/experimental/gluon/language/amd/cdna4/async_copy.py

```diff
@@ -0,0 +1,170 @@
+        cache_modifier (str): Cache modifier specifier. Defaults to "".
+    """
+    _check(isinstance(offsets.type.layout, (BlockedLayout, SliceLayout)),
+           lambda: "expected offsets type layout to be BlockedLayout or SliceLayout")
+    _verify_buffer_ops(ptr, offsets, mask, other)
+
+    mask = _unwrap_if_constexpr(mask)
+    if mask is not None:
+        offsets, mask = _semantic.broadcast_impl_value(offsets, mask)
+    other = _unwrap_if_constexpr(other)
+    if other is not None:
+        offsets, other = _semantic.broadcast_impl_value(offsets, other)
+
+    mask = mask.handle if mask is not None else ir.value()
+    other = other.handle if other is not None else ir.value()
+    stride = ir.value()
+    cache_modifier = _semantic._str_to_load_cache_modifier(cache_modifier)
+
+    _semantic.builder.create_buffer_load_to_local(dest.handle, ptr.handle, offsets.handle, mask, other, stride,
+                                                  cache_modifier)
+
+
+@builtin
+def async_wait(num_outstanding=0, _semantic=None):
+    """
+    Wait for outstanding asynchronous memory operations, this includes
+    normal load like `load` and `buffer_load`, as well as all async memory
+    operations like  `global_load_to_shared` and `buffer_load_to_shared`.
+    It will block until the number of outstanding operations is less than or
+    equal to `num_outstanding`.
+
+    Args:
+        num_outstanding (int): The number of outstanding operations to wait for. Defaults to 0.
+    """
+    num_outstanding = _unwrap_if_constexpr(num_outstanding)
+    _semantic.builder.create_async_wait_group(num_outstanding)
+
+
+@builtin
+def async_hint_shared(smem, _semantic=None):
+    """
+    Gives hints to the underlying compiler to avoid emitting unnecessary waits
+    before loading from the target shared memory for better optimization.
+
+    Args:
+        smem (shared_memory_descriptor): The target shared memory descriptor.
+    """
+    return async_shared_memory_descriptor(smem)
+
+
+class async_shared_memory_descriptor(shared_memory_descriptor):
+    """
+    A wrapper for shared memory descriptor that gives hints to the underlying
+    compiler to avoid emitting unnecessary waits before loading from the
+    shared memory for better optimization.
+    """
+    SYNCED_VIA_WAIT_ATTR_NAME = "ttg.amdgpu.syncedViaAsyncWait"
+
+    def __init__(self, smem):
+        self.handle = smem.handle
+        self.type = smem.type
+
+    @builtin
+    def load(self, layout, _semantic):
+        """
+        Load a tensor from shared memory.
+
+        Args:
+            layout (DistributedLayout): The destination layout of the tensor.
+
+        Returns:
+            tensor: A Gluon tensor containing the loaded data.
+        """
+        layout = _unwrap_if_constexpr(layout)
+        ret = _semantic.shared_load(self, layout)
+        ret.handle.set_attr(self.SYNCED_VIA_WAIT_ATTR_NAME, _semantic.builder.get_bool_attr(True))
```

Comment:
This code is wrong as you haven't defined a custom type for your value, so it won't be reconstructed the same after control flow. e.g.:
```python
smem = async_hint_shared(smem)
for i in range(n):
    # smem is a plain shared_memory_descriptor here, so not loaded with hint
    smem.load(...)
```

tbh though the API seems strange to me. Why not have a custom function instead:
```python
val = async_copy.load_shared_relaxed(smem, layout)
```